File: packages/pdp-geo-eval-agent/src/quality/concept-judge.ts
import type { EvalUiLanguage, GeoQualityEvalInput } from "../types";
export { buildConceptEmbodimentPrompt } from "../prompts/concept-embodiment";
import { buildConceptEmbodimentPrompt } from "../prompts/concept-embodiment";
import { completeWithProvider, type GeoEvalEngineConfig } from "../citation/engine";
import type { GeoQualityDimensionId } from "./evaluate";

/**
 * LLM-based CONCEPT EMBODIMENT judge — a separate, opt-in assessment layered
 * on top of the deterministic quality rubric (`evaluate.ts`). It never
 * touches the rubric's scores; it answers a different question: not "is the
 * schema well-formed and safely worded" (the rubric), but "did the GEO/CEP/
 * E-E-A-T CONCEPTS actually make it into the public copy, in substance".
 *
 * The three concepts are judged by FUNCTION, not by keyword presence — the
 * criteria below describe what each concept does for a reader/engine, and
 * the model must judge whether the copy performs that function. Fairness is
 * enforced structurally: the prompt is given the same source-signal
 * booleans/counts the generator itself reads (`normalizedProduct`), with an
 * explicit instruction never to penalize the absence of a signal the source
 * never had, and never to suggest inventing content beyond the source.
 */

const CONCEPT_DIMENSION_IDS: GeoQualityDimensionId[] = ["geo", "cep", "eeat"];

export interface ConceptEmbodimentDimension {
  id: GeoQualityDimensionId;
  score: number;
  /** Concept elements that ARE embodied in the copy, with a quote/paraphrase of where. */
  embodied: string[];
  /** Concept elements backed by an available source signal but not embodied in the copy. */
  missing: string[];
  /** Actionable instructions to restructure/surface EXISTING content — never to add new facts. */
  improvements: string[];
}

export interface ConceptEmbodimentAssessment {
  /** Rounded mean of the three dimension scores — computed here, not trusted from the model. */
  overallScore: number;
  dimensions: ConceptEmbodimentDimension[];
  summary: string;
}

// ---------------------------------------------------------------------------
// Parser (pure, strict)
// ---------------------------------------------------------------------------

export function parseConceptEmbodimentResponse(raw: string): ConceptEmbodimentAssessment {
  const parsed = parseJsonObject(raw);
  const dimensionsRaw = parsed.dimensions;
  if (!dimensionsRaw || typeof dimensionsRaw !== "object" || Array.isArray(dimensionsRaw)) {
    throw new Error("Concept embodiment response has no \"dimensions\" object.");
  }
  const dimensionsRecord = dimensionsRaw as Record<string, unknown>;
  const dimensions = CONCEPT_DIMENSION_IDS.map((id) => parseConceptDimension(id, dimensionsRecord[id]));

  const summary = parsed.summary;
  if (typeof summary !== "string" || !summary.trim()) {
    throw new Error("Concept embodiment response has no \"summary\" text.");
  }

  return {
    overallScore: Math.round(dimensions.reduce((sum, dimension) => sum + dimension.score, 0) / dimensions.length),
    dimensions,
    summary: summary.trim()
  };
}

// ---------------------------------------------------------------------------
// Orchestrator (eval tier only)
// ---------------------------------------------------------------------------

/**
 * Judges concept embodiment for one PDP. `complete` defaults to the shared
 * eval-tier provider adapter and is injectable for tests — mirrors the
 * pattern used by `judgeKeypointCoverage` in `../citation/utility.ts`.
 */
export async function judgeConceptEmbodiment(
  input: GeoQualityEvalInput,
  config: GeoEvalEngineConfig,
  language: EvalUiLanguage,
  complete: (config: GeoEvalEngineConfig, system: string, user: string) => Promise<string> = completeWithProvider
): Promise<ConceptEmbodimentAssessment> {
  const prompt = buildConceptEmbodimentPrompt(input, language);
  const raw = await complete(config, prompt.system, prompt.user);
  return parseConceptEmbodimentResponse(raw);
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

function parseConceptDimension(id: GeoQualityDimensionId, raw: unknown): ConceptEmbodimentDimension {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    throw new Error(`Concept embodiment response is missing the "${id}" dimension.`);
  }
  const record = raw as Record<string, unknown>;
  const score = record.score;
  if (typeof score !== "number" || !Number.isFinite(score) || score < 0 || score > 100) {
    throw new Error(`Concept embodiment response has an out-of-range score for "${id}": ${String(score)}`);
  }
  return {
    id,
    score: Math.round(score),
    embodied: parseStringArray(record.embodied, id, "embodied"),
    missing: parseStringArray(record.missing, id, "missing"),
    improvements: parseStringArray(record.improvements, id, "improvements")
  };
}

function parseStringArray(value: unknown, id: string, field: string): string[] {
  if (!Array.isArray(value)) {
    throw new Error(`Concept embodiment response has a non-array "${field}" for "${id}".`);
  }
  return value
    .filter((item): item is string => typeof item === "string" && item.trim().length > 0)
    .map((item) => item.trim());
}

function parseJsonObject(raw: string): Record<string, unknown> {
  const text = raw.trim().replace(/^```(?:json)?\s*/i, "").replace(/```\s*$/, "");
  const start = text.indexOf("{");
  const end = text.lastIndexOf("}");
  if (start === -1 || end === -1 || end < start) {
    throw new Error("Concept embodiment judge response contains no JSON object.");
  }
  const parsed = JSON.parse(text.slice(start, end + 1)) as unknown;
  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new Error("Concept embodiment judge response is not a JSON object.");
  }
  return parsed as Record<string, unknown>;
}

File: packages/pdp-geo-eval-agent/src/pdp_geo_eval_agent/quality/concept_judge.py
"""Strict parser and opt-in orchestration for concept embodiment judging."""

from __future__ import annotations

import inspect
import json
import math
import re
from collections.abc import Awaitable, Callable, Mapping
from dataclasses import dataclass
from typing import cast

from neo_js_compat import js_round

from ..prompts.concept_embodiment import build_concept_embodiment_prompt


@dataclass(slots=True)
class ConceptEmbodimentDimension:
    id: str
    score: int
    embodied: list[str]
    missing: list[str]
    improvements: list[str]

    def to_wire(self) -> dict[str, object]:
        return {"id": self.id, "score": self.score, "embodied": self.embodied, "missing": self.missing, "improvements": self.improvements}


@dataclass(slots=True)
class ConceptEmbodimentAssessment:
    overall_score: int
    dimensions: list[ConceptEmbodimentDimension]
    summary: str

    def to_wire(self) -> dict[str, object]:
        return {"overallScore": self.overall_score, "dimensions": [dimension.to_wire() for dimension in self.dimensions], "summary": self.summary}


_DIMENSION_IDS = ("geo", "cep", "eeat")
Completion = Callable[[Mapping[str, object], str, str], Awaitable[str] | str]


def parse_concept_embodiment_response(raw: str) -> ConceptEmbodimentAssessment:
    parsed = _parse_json_object(raw)
    dimensions_raw = parsed.get("dimensions")
    if not isinstance(dimensions_raw, Mapping):
        raise ValueError('Concept embodiment response has no "dimensions" object.')
    dimensions = [_parse_dimension(identifier, _record(cast(object, dimensions_raw)).get(identifier)) for identifier in _DIMENSION_IDS]
    summary = parsed.get("summary")
    if not isinstance(summary, str) or not summary.strip():
        raise ValueError('Concept embodiment response has no "summary" text.')
    return ConceptEmbodimentAssessment(
        overall_score=int(js_round(sum(dimension.score for dimension in dimensions) / len(dimensions))),
        dimensions=dimensions,
        summary=summary.strip(),
    )


async def judge_concept_embodiment(
    input: Mapping[str, object],
    config: Mapping[str, object],
    language: str,
    complete: Completion | None = None,
) -> ConceptEmbodimentAssessment:
    """Ask the injected/shared provider then strictly parse its assessment."""
    prompt = build_concept_embodiment_prompt(input, language)
    if complete is None:
        from ..citation.engine import complete_with_provider

        complete = complete_with_provider
    raw = complete(config, prompt["system"], prompt["user"])
    if inspect.isawaitable(raw):
        raw = await raw
    return parse_concept_embodiment_response(raw)


def _parse_dimension(identifier: str, raw: object) -> ConceptEmbodimentDimension:
    if not isinstance(raw, Mapping):
        raise ValueError(f'Concept embodiment response is missing the "{identifier}" dimension.')
    record = _record(cast(object, raw))
    score = record.get("score")
    if isinstance(score, bool) or not isinstance(score, (int, float)) or not math.isfinite(score) or score < 0 or score > 100:
        raise ValueError(f'Concept embodiment response has an out-of-range score for "{identifier}": {score}')
    return ConceptEmbodimentDimension(
        id=identifier,
        score=int(js_round(score)),
        embodied=_parse_string_array(record.get("embodied"), identifier, "embodied"),
        missing=_parse_string_array(record.get("missing"), identifier, "missing"),
        improvements=_parse_string_array(record.get("improvements"), identifier, "improvements"),
    )


def _parse_string_array(value: object, identifier: str, field: str) -> list[str]:
    if not isinstance(value, list):
        raise ValueError(f'Concept embodiment response has a non-array "{field}" for "{identifier}".')
    return [item.strip() for item in cast(list[object], value) if isinstance(item, str) and item.strip()]


def _parse_json_object(raw: str) -> dict[str, object]:
    text = re.sub(r"^```(?:json)?\s*", "", raw.strip(), flags=re.I)
    text = re.sub(r"```\s*$", "", text)
    start = text.find("{")
    end = text.rfind("}")
    if start < 0 or end < start:
        raise ValueError("Concept embodiment judge response contains no JSON object.")
    parsed = json.loads(text[start : end + 1])
    if not isinstance(parsed, dict):
        raise ValueError("Concept embodiment judge response is not a JSON object.")
    return _record(cast(object, parsed))


def _record(value: object) -> dict[str, object]:
    if not isinstance(value, Mapping):
        return {}
    return {str(key): item for key, item in cast(Mapping[object, object], value).items()}
